<a href="https://www.kaggle.com/code/prasadposture121/analysis-of-data-science-jobs?scriptVersionId=131317666" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# Analysis of Data Science Jobs

# Introduction
The amount of data we have been generating from the past few years has been incresed exponetially. As John Naisbitt once said, "We are drowning in information but starving for knowledge." But to conquer this, there has been a rapid development in the field of data science. This field makes use of mainly statistics, mathematics, scientific methods, computer programming, machine learning, deep learning and many more things to extract knowledge, generate insights and derive patterns from the noisy, unstrutured data. The growth of this field has brought many job opportunites across the world, due to this, the data science has become the most sought after career option today. This field will continue to grow, since the need of analayzing the mammoth  amount of data is increasing day-by-day. There are various job titles among this field and to understand about them, here I have data of 3755 job titles from the companies all across the world. I will perform question based analysis to generate various insights and to see what are the main factors that affect the relation of job titles and their salaries. If you find this notebook useful please upvote and if you have any sugesstions or queries please feel free to contact me. Thank you.

# Importing Dependencies

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')
import plotly.express as px
import plotly.graph_objects as go

# Loading the Data

In [2]:
df = pd.read_csv('/kaggle/input/data-science-salaries-2023/ds_salaries.csv')
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [3]:
# Shape of the dataset i.e. No of rows and columns
df.shape

(3755, 11)

In [4]:
# General Information of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


In [5]:
# Statistical Information of the Numeric Attributes
df.describe()

work_year        salary  salary_in_usd  remote_ratio
count  3755.000000  3.755000e+03    3755.000000   3755.000000
mean   2022.373635  1.906956e+05  137570.389880     46.271638
std       0.691448  6.716765e+05   63055.625278     48.589050
min    2020.000000  6.000000e+03    5132.000000      0.000000
25%    2022.000000  1.000000e+05   95000.000000      0.000000
50%    2022.000000  1.380000e+05  135000.000000      0.000000
75%    2023.000000  1.800000e+05  175000.000000    100.000000
max    2023.000000  3.040000e+07  450000.000000    100.000000

In [6]:
# Statistical Information of the Categorical Attributes
df.describe(include=['O'])

experience_level employment_type      job_title salary_currency  \
count              3755            3755           3755            3755   
unique                4               4             93              20   
top                  SE              FT  Data Engineer             USD   
freq               2516            3718           1040            3224   

       employee_residence company_location company_size  
count                3755             3755         3755  
unique                 78               72            3  
top                    US               US            M  
freq                 3004             3040         3153

In [7]:
# Checking for the missing values
df.isna().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

# Data Manipulation and Feature Engineering

### Converting Country Codes (ISO3)

In [8]:
! pip install country_converter

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.5/44.5 kB 3.1 MB/s eta 0:00:00


In [9]:
# Converting country codes of employee_residence column
import country_converter as coco
df['employee_residence'] = coco.convert(names=df['employee_residence'])
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [10]:
# Converting country codes of company_location column
import country_converter as coco
df['company_location'] = coco.convert(names=df['company_location'])
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size  
0              ESP            L  
1              USA            S  
2              USA            S  
3              CAN            M  
4              CAN            M

### Foreign Employees
Adding a new column based on the condition that whether the country of residence of employee and the country where the company is based are same or not. If same then the employee is not a foreign employee if not then he/she is a foreign employee.

In [11]:
df['foreign_employee']=np.where(df['employee_residence']==df['company_location'],"No","Yes")
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size foreign_employee  
0              ESP            L               No  
1              USA            S               No  
2              USA            S               No  
3              CAN            M               No  
4              CAN            M               No

### Currency Conversion Rates
We have been given the salaries in different currerncies, also there is another column which contains the salaries in USD. We can use the information to determine the conversion rates between the different currencies. We will use USD as a standard currency here.

In [12]:
df['conversion_rates'] = df['salary']/df['salary_in_usd']
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size foreign_employee  conversion_rates  
0              ESP            L               No           0.93189  
1              USA            S               No           1.00000  
2              USA            S               No           1.00000  
3              CAN            M               No           1.00000  
4              CAN            M               No           1.00000

### Experience Level
It is given in short format, we will elaborate it using replace method.

In [13]:
df['experience_level'].replace(['SE', 'MI', 'EN', 'EX'],
                               ['Senior-level','Mid-level','Entry-level','Executive-level'],
                               inplace=True)
df.head()

work_year experience_level employment_type                 job_title  \
0       2023     Senior-level              FT  Principal Data Scientist   
1       2023        Mid-level              CT               ML Engineer   
2       2023        Mid-level              CT               ML Engineer   
3       2023     Senior-level              FT            Data Scientist   
4       2023     Senior-level              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size foreign_employee  conversion_rates  
0              ESP            L               No           0.93189  
1              USA            S               No           1.00000  
2              USA            S               No           1.00000  
3              CAN            M               No           1.00000  
4              CAN            M               No           1.00000

### Employment Type
We will also elaborate this one

In [14]:
df['employment_type'].replace(['FT', 'CT', 'FL', 'PT'],['Full-time','Contract','Freelancer','Part-time'], inplace=True)
df.head()

work_year experience_level employment_type                 job_title  \
0       2023     Senior-level       Full-time  Principal Data Scientist   
1       2023        Mid-level        Contract               ML Engineer   
2       2023        Mid-level        Contract               ML Engineer   
3       2023     Senior-level       Full-time            Data Scientist   
4       2023     Senior-level       Full-time            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size foreign_employee  conversion_rates  
0              ESP            L               No           0.93189  
1              USA            S               No           1.00000  
2              USA            S               No           1.00000  
3              CAN            M               No           1.00000  
4              CAN            M               No           1.00000

### Job Type
Using the remote_ratio, we will replace the values with remote(100%), on-site(0%) and hybrid(50%).

In [15]:
df['remote_ratio'].replace([0,50,100],['On-Site','Hybrid','Remote'], inplace=True)
df.rename(columns={'remote_ratio':'job_type'}, inplace=True) #Renaming the column
df.head()

work_year experience_level employment_type                 job_title  \
0       2023     Senior-level       Full-time  Principal Data Scientist   
1       2023        Mid-level        Contract               ML Engineer   
2       2023        Mid-level        Contract               ML Engineer   
3       2023     Senior-level       Full-time            Data Scientist   
4       2023     Senior-level       Full-time            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence job_type  \
0   80000             EUR          85847                ESP   Remote   
1   30000             USD          30000                USA   Remote   
2   25500             USD          25500                USA   Remote   
3  175000             USD         175000                CAN   Remote   
4  120000             USD         120000                CAN   Remote   

  company_location company_size foreign_employee  conversion_rates  
0              ESP            L               No           0.93189  
1              USA            S               No           1.00000  
2              USA            S               No           1.00000  
3              CAN            M               No           1.00000  
4              CAN            M               No           1.00000

# Question Based Analysis

<b> 1) Which all countries took part in this survey?

In [16]:
fig = px.choropleth(locations = df['employee_residence'],
                    color = df['employee_residence'],
                    title = 'Countries involved in the Survey')
fig.show()

<b> 2) Which are the top 10 popular job designations?

In [17]:
df['job_title'].value_counts().head(10)

Data Engineer                1040
Data Scientist                840
Data Analyst                  612
Machine Learning Engineer     289
Analytics Engineer            103
Data Architect                101
Research Scientist             82
Data Science Manager           58
Applied Scientist              58
Research Engineer              37
Name: job_title, dtype: int64

In [18]:
px.bar(x=df['job_title'].value_counts().head(10).index,
       y=df['job_title'].value_counts().head(10),
       title='Top 10 Most Popular Job Designations',
      labels={'y':'No. of posts','x':'Job Designations'})

Data Engineer is one of the most popular job designation followed by Data Scientist then Data Analyst.

<b> 3) Which are the top 10 highest paid job designations over the years?

In [19]:
xdf=df.groupby(['job_title'])['salary_in_usd'].median().sort_values(ascending=False).head(10)
xdf

job_title
Data Science Tech Lead                 375000.0
Cloud Data Architect                   250000.0
Data Lead                              212500.0
Data Analytics Lead                    211254.5
Head of Data                           202500.0
Principal Data Engineer                192500.0
Applied Scientist                      191737.5
Principal Machine Learning Engineer    190000.0
Data Science Manager                   183780.0
Data Infrastructure Engineer           183655.0
Name: salary_in_usd, dtype: float64

In [20]:
px.bar(x=xdf.index, y=xdf, title='Top 10 High Paying Job Designations',
      labels={'y':'Median Salary','x':'Job Designations'})

<b> 4) Compare the company sizes and salaries offered by them over the years.

In [21]:
table = pd.pivot_table(df,values='salary_in_usd',index='work_year',columns='company_size', aggfunc='median')
table

company_size         L         M        S
work_year                                
2020           85000.0   96500.0  51321.0
2021           90734.0   64411.5  76833.0
2022          120000.0  135000.0  58418.5
2023          123700.0  145000.0  74875.5

In [22]:
px.box(df, x='company_size', y='salary_in_usd', color='work_year',
       category_orders={"company_size":["S","M","L"],"work_year":[2020, 2021, 2022, 2023]},
      title='Company Size vs Salary')

The medium sized companies have the highest median salaries over the years also show wide distribution for the year 2022 and 2023. The large sized companies follow a trend of increased median salaries over the years. On the small sized companies show flactuation in their salaries.

<b> 5) Which companies are preferred by employees if we consider their sizes?

In [23]:
px.pie(df,names=df['company_size'].value_counts().index, values=df['company_size'].value_counts(),
      title='Proportion of Company Size')

84% employees prefer to work in a medium size company the reason behind this could be the higher median salary.

<b> 6) Give the number of employees working remote, on-site, hybrid every year?

In [24]:
table = df.pivot_table(index='work_year',values='salary',columns='job_type',aggfunc='count')
table

job_type   Hybrid  On-Site  Remote
work_year                         
2020           21       16      39
2021           76       34     120
2022           62      711     891
2023           30     1162     593

In [25]:
px.histogram(x=df['work_year'],color=df['job_type'],barmode='group',labels={'x':'Work Year'},title='Count of Job Types wih Year')

The number of onsite jobs has exponetially increased from 2020 to 2023. The number of remote job saw increased form 2020 to 2022 then they got decreased in 2023. Hybrid jobs didn't show much variation.

<b> 7) Which job type get higher salaries over the year?

In [26]:
table=df.pivot_table(values='salary_in_usd', index='work_year', columns='job_type')
table

job_type         Hybrid        On-Site         Remote
work_year                                            
2020       77591.238095   93426.187500   99763.205128
2021       76104.407895   92899.676471  105812.783333
2022       84560.870968  134719.236287  135631.104377
2023       72053.500000  152393.521515  146380.124789

In [27]:
px.box(df,x='work_year', y='salary_in_usd', color='job_type',
      title='Distribution of Salaries over the Job Types with Year')

In year 2020 due to pandemic many people were working either remote or hybrid hence the median salary for both of these job-types is higher than the on-site. The medan salary of on-site workers employees increases gradually and is nearly equal to the remote workers in 2022 and overtakes them by smaller margin in 2023. Both on-site and remote show wide distribution of salaries, while there isn't much variation in the salaries of people working in hybrid mode.

<b> 8) What percentage of remote workers are foreign?

In [28]:
df['foreign_employee'].value_counts()/len(df['foreign_employee'])

No     0.974434
Yes    0.025566
Name: foreign_employee, dtype: float64

In [29]:
px.pie(values=df['foreign_employee'].value_counts(),
       names=df['foreign_employee'].value_counts().index,
      title='Proportion of Foreign Employees')

Only 2.56% of the total people surveyed are foreign employees.

<b> 9) In which country most foreign employees work?

In [30]:
xdf=df[df['foreign_employee']=='Yes']['company_location'].value_counts()
xdf.head(10)

USA    41
DEU    10
GBR     6
AUS     5
CAN     4
SGP     2
LUX     2
FRA     2
ASM     2
ESP     2
Name: company_location, dtype: int64

In [31]:
fig = px.choropleth(locations = xdf.index, color = xdf, title = 'Countries with Remote Foreign Employees')
fig.show()

Most foreing employees work in USA followed by Germany.

<b> 10) Which country has the most number of people working for foreign companies?

In [32]:
xdf=df[df['foreign_employee']=='Yes']['employee_residence'].value_counts()
xdf.head(10)

IND    15
FRA     6
USA     5
PAK     5
ESP     5
ITA     4
PRT     4
BRA     3
ARG     3
POL     2
Name: employee_residence, dtype: int64

In [33]:
fig = px.choropleth(locations = xdf.index, color = xdf, title = 'Countries with People working in Foreign Country')
fig.show()

India tops the list of No. of people working for a foreign country followed by France.

<b> 11) Which currency is the weakest and which one is the strongest? See their variations.</b><br>
We can compare the currency conversion rates while using USD as the standard currency and determine the strenths of the currencies,

In [34]:
curr_stgh=df.groupby(['salary_currency'])['conversion_rates'].median()

In [35]:
# Top 5 Strongest Currencies
curr_stgh.sort_values().head(5)

salary_currency
GBP    0.812134
EUR    0.931899
CHF    0.955135
USD    1.000000
CAD    1.301833
Name: conversion_rates, dtype: float64

In [36]:
xdf=df[df.conversion_rates<1.36]
px.box(xdf, x='salary_currency',y='conversion_rates')

Since we have taken USD as the standard currency it doesn't show any variation. On the other hand EUR and CAD show much variation. And GBP continues to be storngest currency throughout.

In [37]:
# Top 5 Weakest Currencies
curr_stgh.sort_values(ascending=False).head(5)

salary_currency
CLP    759.278685
HUF    307.821464
JPY    109.870224
INR     78.627701
THB     33.952756
Name: conversion_rates, dtype: float64

In [38]:
xdf=df[df.conversion_rates>30.0]
px.box(xdf, x='salary_currency',y='conversion_rates')

CLP remains weakest of all with no much variations. HUF shows large variations in its strength.

<b>Work in Progress